odl/trafos/convolution.py

```diff
@@ -0,0 +1,849 @@
+            ``kernel`` is callable.
+
+        Returns
+        -------
+        kernel : `DiscreteLpElement`
+            Element in an adequate space.
+        range : `DiscreteLp`
+            Range of the convolution operator.
+        """
+        if isinstance(kernel, domain.element_type):
+            if kernel.ndim != domain.ndim:
+                raise ValueError('`kernel` has dimension {}, expected {}'
+                                 ''.format(kernel.ndim, domain.ndim))
+            # Set the range equal to the domain shifted by the midpoint of
+            # the kernel space.
+            kernel_shift = kernel.space.partition.mid_pt
+            range = uniform_discr_fromdiscr(
+                domain, min_pt=domain.min_pt + kernel_shift)
+
+        elif callable(kernel):
+            # Use "natural" kernel space, which is the zero-centered
+            # version of the domain.
+            extent = domain.partition.extent()
+            std_kernel_space = uniform_discr_fromdiscr(
+                domain, min_pt=-extent / 2)
+            kernel = std_kernel_space.element(kernel, **kernel_kwargs)
+            range = domain
+
+        else:
+            # Make a zero-centered space with the same cell sides as
+            # domain, but shape according to the kernel.
+            kernel = np.asarray(kernel, dtype=domain.dtype)
+            shape = kernel.shape
+            extent = domain.cell_sides * shape
+            kernel_space = uniform_discr_fromdiscr(
+                domain, min_pt=-extent / 2, shape=shape)
+            kernel = kernel_space.element(kernel)
+            range = domain
+
+        return kernel, range
+
+    @property
+    def impl(self):
+        """Implementation of this operator."""
+        return self.__impl
+
+    def kernel(self):
+        """Return the real-space kernel of this convolution operator.
+
+        Note that internally, the scaled kernel is stored if
+        ``scale=True`` (default) was chosen during initialization, i.e.
+        calculating the original kernel requires to reverse the scaling.
+        Otherwise, the original unscaled kernel is returned.
+        """
+        if self._scale:
+            return self.__kernel / self.__kernel_scaling
+        else:
+            return self.__kernel
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        # No resampling needed
+        return signal.convolve(x, self.__kernel, mode='same')
+
+    @property
+    def adjoint(self):
+        """Adjoint operator defined by the reversed kernel."""
+        # Make space for the adjoint kernel, -S if S is the kernel space
+        adj_min_pt = -self._kernel.space.max_pt
+        adj_max_pt = -self._kernel.space.min_pt
+        adj_ker_space = uniform_discr_fromdiscr(
+            self._kernel.space,
+            min_pt=adj_min_pt, max_pt=adj_max_pt)
+
+        # Adjoint kernel is k_adj(x) = k(-x), i.e. the kernel array is
```

Comment:
This computation method only holds if the domain is centered at 0, right? Otherwise you need a more precise method.